### DISCUSSION TITLE
Download audio loaded in wavesurfer 7 with the envelope plugin effects

### DISCUSSION THREAD

**Discussion by Nosa-Drexx** [katspaugh/wavesurfer.js#3227] [Q&A]
I'm new to wavesurfer and I'm trying to download an audio that has already been loaded into wavesurfer and has been affected by the wavesurfer envelope plugin.
.
I want to download that audio with the effects from the wavesurfer envelope plugin.
I've tried the `getDecodedData` method, but after trigging a download of the buffer returned It seems to only download audio without the effect from the envelope plugin.
.
I'm currently using the version 7 of the wavesurfer.js package 

Would greatly appreciate the help from anyone 🙏

**Comment by katspaugh** [comment]
You cannot download a post-envelope audio because envelope just changes the volume in run time, it doesn’t change the underlying data.

**Reply by Nosa-Drexx** [reply]
Thanks for the feedback @katspaugh 

Though is there a way I could change the underlying data with the data from the envelope plugin

**Reply by katspaugh** [reply]
You can theoretically do that by multiplying each float in the decoded data by the corresponding volume at that position.

**Reply by Nosa-Drexx** [reply]
Alright thanks @katspaugh.

**Reply by Nosa-Drexx** [reply]
Any help with how I can get the floats of the audio data @katspaugh

**Reply by katspaugh** [reply]
getDecodedData

**Reply by Nosa-Drexx** [reply]
Alright thanks @katspaugh

**Reply by Nosa-Drexx** [reply]
@katspaugh thanks for your help, I've been able to create a function that takes in the points from wavesurfer envelope plugin using the `getPoint` method, clones the  wavesurfer buffer, and mutates the cloned buffer with new effects of the envelope so it can be downloaded with the post-envelope effect

**Reply by katspaugh** [reply]
Nice! If you’re planning to play the resulting audio, you might want to pass a higher [sampleRate option](https://wavesurfer.xyz/docs/types/wavesurfer.WaveSurferOptions) like 44100 because by default it’s the lowest possible, 8000.

**Reply by mkhaziq** [reply]
@Nosa-Drexx can you share the function that you have created to download the updated audio. As I am also new to wavesurfer, need help with it.

**Reply by zenozaga** [reply]
I have created this function for my personal use; I hope it is helpful.

` 

    /**
     * Applies volume changes to an audio buffer based on specified points.
     *
     * @param {AudioBuffer} audioBuffer - The input audio buffer.
     * @param {Array<{volume: number, time: number}>} points - An array of objects representing volume/time points.
     * @param {AudioContext|OfflineAudioContext|undefined} audioCtx - The audio context to use for rendering.
     * @returns {Promise<AudioBuffer>} - A promise that resolves with the modified audio buffer.
     */
    async function volumeLevels(audioBuffer, points = [], audioCtx) {


        // Sort points by time
        points.sort((a, b) => a.time - b.time);

        // Create an offline audio context with the same number of channels, length, and sample rate as the input audio buffer
        if(!audioCtx) audioCtx = new OfflineAudioContext(audioBuffer.numberOfChannels, audioBuffer.length, audioBuffer.sampleRate);

        // Create a gain node
        var gainNode = audioCtx.createGain();

        // Create a buffer source node
        var source = audioCtx.createBufferSource();
        source.buffer = audioBuffer;

        // Connect the source to the gain node and the gain node to the audio context destination
        source.connect(gainNode);
        gainNode.connect(audioCtx.destination);

        // Set the initial gain to 0 at the current time
        const currentTime = audioCtx.currentTime;
        gainNode.gain.setValueAtTime(0, currentTime);

        // Apply volume changes based on the specified points
        points.forEach((point, index) => {

            const nextPoint = points[index + 1];

            if (nextPoint) {
                // If there is a next point, linearly ramp the gain from the current point to the next point
                const duration = nextPoint.time - point.time;
                gainNode.gain.linearRampToValueAtTime(point.volume, currentTime + point.time);
                gainNode.gain.linearRampToValueAtTime(nextPoint.volume, currentTime + point.time + duration);
            } else {
                // If it's the last point, sustain the volume
                gainNode.gain.linearRampToValueAtTime(point.volume, currentTime + point.time);
            }

        });

        // Start playing the source in the offline audio context
        await source.start();

        if(audioCtx instanceof OfflineAudioContext){
            // Render the audio context
            return await audioCtx.startRendering();;
        }
        
        return null;

        
    }

    
    //  getting points from envelope plugin
    const points= envelope.getPoints();
    
    // If you wish to listen to the result, you can pass an AudioContext as an argument. 
    // If audioCtx is not defined, the AudioBuffer will be returned.
    var buffer = await volumeLevels(ws.getDecodedData(), volumePoints, new AudioContext())

 `
